pkg/gatewayserver/gatewayserver.go

```diff
@@ -638,6 +644,76 @@ func (gs *GatewayServer) handleUpstream(conn connectionEntry) {
 	}
 }
 
+func (gs *GatewayServer) handleLocationUpdates(conn connectionEntry) {
+	ctx := conn.Context()
+	uid := unique.ID(ctx, conn.Gateway())
+
+	logger := log.FromContext(ctx).WithFields(log.Fields(
+		"protocol", conn.Frontend().Protocol(),
+		"gateway_uid", uid,
+	))
+
+	var err error
+	var callOpt grpc.CallOption
+	callOpt, err = rpcmetadata.WithForwardedAuth(ctx, gs.AllowInsecureForCredentials())
+	if errors.IsUnauthenticated(err) {
+		callOpt = gs.WithClusterAuth()
+	} else if err != nil {
+		return
+	}
+	registry, err := gs.getRegistry(ctx, &conn.Gateway().GatewayIdentifiers)
+	if err != nil {
+		return
+	}
+
+	var location *ttnpb.Location
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case location = <-conn.Location():
+			// Drop older updates
+			for len(conn.Location()) > 0 {
+				location = <-conn.Location()
+			}
+
+			antennas := conn.Gateway().Antennas
+			if location != nil && len(antennas) > 0 {
+				// TODO: Handle multiple antenna locations (https://github.com/TheThingsNetwork/lorawan-stack/issues/2006).
+				location.Source = ttnpb.SOURCE_GPS
+				antennas[0].Location = *location
+
+				conn.UpdateAntennas(antennas)
+			} else {
+				logger.Warnf("No antenna, not updating location")
+			}
+
+			_, err := registry.Update(ctx, &ttnpb.UpdateGatewayRequest{
+				Gateway: ttnpb.Gateway{
+					GatewayIdentifiers: conn.Gateway().GatewayIdentifiers,
+					Antennas:           conn.Gateway().Antennas,
+				},
+				FieldMask: pbtypes.FieldMask{
+					Paths: []string{
+						"antennas",
+					},
+				},
+			}, callOpt)
+
+			if err != nil {
+				logger.WithError(err).Errorf("Antenna location update failed")
```

Comment:
- Errors are for findings bugs, use `Warn`
- No need for the formatted version
- `Failed to ...`